### PR TITLE
Issue #104: automatically refresh worksheet if some bundle is not ready

### PR DESCRIPTION
**Pull request by ChengshuLi**
@percyliang

### CODE REVIEW

File: codalab/client/local_bundle_client.py [hunk 1]
```diff
@@ -1050,6 +1050,8 @@ def resolve_interpreted_items(self, interpreted_items):
         The result can be serialized via JSON.
         """
         for item in interpreted_items:
+            if item == None:
```
Comment (percyliang) [inline codalab/client/local_bundle_client.py:894]: replace '==' with 'is'

File: codalab/rest/legacy.py [hunk 2]
```diff
@@ -122,6 +123,22 @@ def full_worksheet(self, uuid):
...
             worksheet_info['error'] = str(e)
 
+        if bundle_uuids:
+            for i, item in enumerate(interpreted_items['items']):
+                if ('bundle_info' not in item):
+                    interpreted_items['items'][i] = None
+                else:
+                    if isinstance(item['bundle_info'], dict):
+                        item['bundle_info'] = [item['bundle_info']]
+                    is_relavant_item = False
```
Comment (percyliang) [inline codalab/rest/legacy.py]: spelling: relevant

File: codalab/rest/legacy.py [hunk 3]
```diff
@@ -122,6 +123,22 @@ def full_worksheet(self, uuid):
             interpreted_items = {'items': []}
             worksheet_info['error'] = str(e)
 
+        if bundle_uuids:
+            for i, item in enumerate(interpreted_items['items']):
+                if ('bundle_info' not in item):
```
Comment (percyliang) [inline codalab/rest/legacy.py]: don't need parens

File: codalab/rest/legacy.py [hunk 4]
```diff
@@ -122,6 +123,22 @@ def full_worksheet(self, uuid):
             interpreted_items = {'items': []}
             worksheet_info['error'] = str(e)
 
+        if bundle_uuids:
```
Comment (percyliang) [inline codalab/rest/legacy.py:169]: Need high-level documentation on what's going on here

File: codalab/rest/legacy.py [hunk 5]
```diff
@@ -135,6 +152,8 @@ def decode_lines(interpreted):
 
         # Currently, only certain fields are base64 encoded.
         for item in worksheet_info['items']:
+            if item == None:
```
Comment (percyliang) [inline codalab/rest/legacy.py]: is

File: codalab/rest/legacy.py [hunk 6]
```diff
@@ -149,6 +168,8 @@ def decode_lines(interpreted):
                 elif isinstance(item['bundle_info'], dict):
                     infos = [item['bundle_info']]
                 for bundle_info in infos:
+                    if bundle_info == None:
```
Comment (percyliang) [inline codalab/rest/legacy.py]: is

File: codalab/rest/legacy.py [hunk 7]
```diff
@@ -447,6 +469,23 @@ def get_worksheet_content(uuid):
     service = BundleService()
     return service.full_worksheet(uuid)
 
+# Needs to store the last unfinished bundle uuids because if a bundle just finished running,
+# unfinished_bundle_uuids won't include that bundle (from the backend points of view, it's a finished bundle)
+# whereas the frontend is still waiting to update its state from running to ready/failed.
+last_unfinished_bundle_uuids = set()
```
Comment (percyliang) [inline codalab/rest/legacy.py]: We should avoid global state like this.  What happens if there are many frontends calling this method...?  The frontend should pass in the set of bundles it think are unfinished.

File: codalab/rest/legacy.py [hunk 8]
```diff
@@ -447,6 +469,23 @@ def get_worksheet_content(uuid):
...
+# Get items that include unfinished run bundles in a worksheet, uuid is the uuid for that worksheet
+@get('/api/worksheets/unfinished_bundles/<uuid:re:%s>/' % spec_util.UUID_STR)
+def get_worksheet_content_with_unfinished_bundles(uuid):
+    global last_unfinished_bundle_uuids
+    service = BundleService()
+    unfinished_bundle_uuids = [bundle.uuid for bundle in local.model.batch_get_bundles(state=State.OPTIONS - State.FINAL_STATES)]
```
Comment (percyliang) [inline codalab/rest/legacy.py]: I think the frontend should just pass in the unfinished bundles and then this method just pass back the info for them.  In fact, this method should have nothing to do with finished/unfinished - all it does is just pass back information for an arbitrary subset of bundles.  Can we just include this list as query params for the existing get_worksheet_content endpoint?

**Comment by percyliang** [comment]
@ChengshuLi Please take another pass

**Comment by ChengshuLi** [comment]
@percyliang

File: codalab/rest/legacy.py [hunk 9]
```diff
@@ -159,20 +159,23 @@ def full_worksheet(self, uuid, bundle_uuids=None):
             interpreted_items = {'items': []}
             worksheet_info['error'] = str(e)
 
+        # bundle_uuids is an optional argument that, if exists, contain the uuids of all the unfinished run bundles that need to be updated
```
Comment (percyliang) [inline codalab/rest/legacy.py]: Document what exactly is returned more precisely.

**Comment by percyliang** [comment]
LGTM modulo comments

**Comment by ChengshuLi** [comment]
@percyliang Feel free to check out this branch to test performance for https://github.com/codalab/codalab-worksheets/pull/169

File: codalab/rest/legacy.py [hunk 10]
```diff
@@ -158,6 +159,27 @@ def full_worksheet(self, uuid):
             interpreted_items = {'items': []}
             worksheet_info['error'] = str(e)
 
+        # bundle_uuids is an optional argument that, if exists, contain the uuids of all the unfinished run bundles that need updating
+        # In this case, full_worksheet will return a list of item parallel to ws.info.items that contain only items that need updating.
+        # More spefically, all items that don't contain run bundles that need updating are None.
```
Comment (percyliang) [inline codalab/rest/legacy.py]: typo

**Comment by ChengshuLi** [comment]
@percyliang

**Comment by percyliang** [comment]
LGTM
But tests are are failing, so please fix before merging.

**Pull request merged**